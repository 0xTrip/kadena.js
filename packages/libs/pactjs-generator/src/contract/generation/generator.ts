import { Arg, Defcap, Defun, Module, Output } from '../parsing/parser';

const keywordsMap: Record<string, string> = {
  decimal: 'number',
  integer: 'number',
  time: 'Date',
  bool: 'boolean',
  guard: 'Function',
  'object{fungible-v2.account-details}': 'object',
};

const mapType = (name: string | undefined): string => {
  if (name === undefined) {
    return 'undefined';
  }
  if (name in keywordsMap) {
    return keywordsMap[name];
  }
  // TODO: add automatic type mapping for other types (like object{fungible-v2....})
  return name;
};

type ModuleName = string;

function capitalize([first, ...rest]: string): string {
  return first.toUpperCase() + rest.join('');
}

/**
 * @alpha
 */
export function generateDts(modules: Output): Map<ModuleName, string> {
  const moduleDtss: Map<ModuleName, string> = new Map<ModuleName, string>();
  for (const ModuleName in modules) {
    if (Object.prototype.hasOwnProperty.call(modules, ModuleName)) {
      const module: Module = modules[ModuleName];
      moduleDtss.set(
        ModuleName,
        `
import type { ICommandBuilder } from '@kadena/client';

declare module '@kadena/client' {
  export type I${capitalize(module.name)}Caps = {
    ${Object.keys(module.defcaps)
      .map((defcapName) => {
        const defcap: Defcap = module.defcaps[defcapName];
        return `"${module.name}.${defcapName}": [ ${Object.keys(defcap.args)
          .map((argName) => {
            const defcapArg: Arg = defcap.args[argName];
            return `${argName.replace(/-/g, '')}: ${mapType(defcapArg.type)}`;
          })
          .join(', ')} ]`;
      })
      .join(', \n')}
  }
  export interface IPactModules {
    "${module.name}": {
       ${Object.keys(module.defuns)
         .map((defun) => {
           const fnc: Defun = module.defuns[defun];
           return `"${fnc.method}": (${Object.keys(fnc.args)
             .map((arg) => {
               const argDef: Arg = fnc.args[arg];
               return `${argDef.name.replace(/-/, '_')}: ${mapType(
                 argDef.type,
               )}`;
             })
             .join(', ')}) => ICommandBuilder<I${capitalize(module.name)}Caps>`;
         })
         .join(',\n')}
    }
  }
}`,
      );
    }
  }
  return moduleDtss;
}
