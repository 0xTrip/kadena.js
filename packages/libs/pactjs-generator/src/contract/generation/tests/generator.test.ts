import { generateDts } from '../generator';
import { StringContractDefinition } from '../StringContractDefinition';

describe('generator', () => {
  it('creates a typescript definition from a contract', () => {
    const contract: string = `(module coin
      (defun transfer:string (from:string to:string amount:decimal))
    )`;
    const parsedContract = new StringContractDefinition(contract);
    const dTs = generateDts(parsedContract.modulesWithFunctions)
      .get('coin')!
      .split(/[\s\n]/)
      .filter((x) => x !== '')
      .join(' ');
    const expected = `import type { ICommandBuilder } from '@kadena/client';
declare module '@kadena/client' {
  export type ICoinCaps = { }
  export interface IPactModules {
    "coin": {
      "transfer": (from: string, to: string, amount: number) => ICommandBuilder<ICoinCaps>
    }
  }
}`
      .split(/[\s\n]/)
      .filter((x) => x !== '')
      .join(' ');
    expect(dTs).toBe(expected);
  });

  it('creates a typescript definition with DEFCAPS from a contract', () => {
    const contract: string = `(module coin
      (defun transfer:string (from:string to:string amount:decimal))
      (defcap GAS ())
      (defcap TRANSFER (sender:string receiver:string amount:decimal))
    )`;
    const parsedContract = new StringContractDefinition(contract);
    const dTs = generateDts(parsedContract.modulesWithFunctions)
      .get('coin')!
      .split(/[\s\n]/)
      .filter((x) => x !== '')
      .join(' ');
    const expected = `import type { ICommandBuilder } from '@kadena/client';
declare module '@kadena/client' {
  export type ICoinCaps = {
    "coin.GAS": [ ],
    "coin.TRANSFER": [ sender: string, receiver: string, amount: number ]
  }
  export interface IPactModules {
    "coin": {
      "transfer": (from: string, to: string, amount: number) => ICommandBuilder<ICoinCaps>
    }
  }
}`
      .split(/[\s\n]/)
      .filter((x) => x !== '')
      .join(' ');
    expect(dTs).toBe(expected);
  });
});
