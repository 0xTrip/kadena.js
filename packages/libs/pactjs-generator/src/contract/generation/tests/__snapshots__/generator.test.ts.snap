// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateDts > add defpact functions to defpact property 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICommonCapabilities {
  (name: 'coin.GAS'): ICap;
  // let users use any other capabilities that can not be inferred
  (name: string, ...args: any[]): ICap;
}

declare module '@kadena/client' {
  interface IPactModules {
    
    \\"user.test-module\\": {
    

      \\"defpact\\":{
        \\"test\\": () => string & { capability : ICommonCapabilities} 
      }
    }
  }
}"
`;

exports[`generateDts > adds IPactReference as the arg type if it is a module reference 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICommonCapabilities {
  (name: 'coin.GAS'): ICap;
  // let users use any other capabilities that can not be inferred
  (name: string, ...args: any[]): ICap;
}

interface ICapability_test_func {
  /**
  * this is defcap doc
  */
  (
    capabilityName: \\"user.test-module.test-cap\\", 
    name: string): ICap,
}

declare module '@kadena/client' {
  interface IPactModules {
    /**
    this is module doc    
    */
    \\"user.test-module\\": {
      /**
      * this is defun doc
      */
      \\"test-func\\": (
        parameterone: PactReference,
        parametertwo: boolean | PactReference) => string & { capability : ICapability_test_func & ICommonCapabilities} 

    }
  }
}"
`;

exports[`generateDts > adds module, capability and function docs to the dts file 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICommonCapabilities {
  (name: 'coin.GAS'): ICap;
  // let users use any other capabilities that can not be inferred
  (name: string, ...args: any[]): ICap;
}

interface ICapability_test_func {
  /**
  * this is defcap doc
  */
  (
    capabilityName: \\"user.test-module.test-cap\\", 
    name: string): ICap,
}

declare module '@kadena/client' {
  interface IPactModules {
    /**
    this is module doc    
    */
    \\"user.test-module\\": {
      /**
      * this is defun doc
      */
      \\"test-func\\": (
        parameterone: object | PactReference,
        parametertwo: boolean | PactReference) => string & { capability : ICapability_test_func & ICommonCapabilities} 

    }
  }
}"
`;

exports[`generateDts > adds some _ to capabilityName to make it unique and avoid name collision if the capability function has also an argument exactly with the same name 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICommonCapabilities {
  (name: 'coin.GAS'): ICap;
  // let users use any other capabilities that can not be inferred
  (name: string, ...args: any[]): ICap;
}

interface ICapability_defpact_test_func {
  /**
  * this is defcap doc
  */
  (
    __capabilityName: \\"user.test-module.test-cap\\", 
    capabilityName: string, 
    _capabilityName: string): ICap,
}

declare module '@kadena/client' {
  interface IPactModules {
    /**
    this is module doc    
    */
    \\"user.test-module\\": {
    

      \\"defpact\\":{
        /**
        * this is defpact doc
        */
        \\"test-func\\": (
          parameterone: any,
          parametertwo: any) => string & { capability : ICapability_defpact_test_func & ICommonCapabilities} 
      }
    }
  }
}"
`;

exports[`generateDts > does not generate capability interface if the function uses no capabilities 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICommonCapabilities {
  (name: 'coin.GAS'): ICap;
  // let users use any other capabilities that can not be inferred
  (name: string, ...args: any[]): ICap;
}

declare module '@kadena/client' {
  interface IPactModules {
    
    \\"user.test-module\\": {
      \\"test-func\\": (
        parameterone: IPactDecimal | PactReference,
        parametertwo: boolean | PactReference) => string & { capability : ICommonCapabilities} 

    }
  }
}"
`;

exports[`generateDts > function parameter is empty if the defun has no param 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICommonCapabilities {
  (name: 'coin.GAS'): ICap;
  // let users use any other capabilities that can not be inferred
  (name: string, ...args: any[]): ICap;
}

declare module '@kadena/client' {
  interface IPactModules {
    
    \\"user.test-module\\": {
      \\"test\\": () => string & { capability : ICommonCapabilities} 

    }
  }
}"
`;

exports[`generateDts > prefixes capabulities for defpact with defpact_ 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICommonCapabilities {
  (name: 'coin.GAS'): ICap;
  // let users use any other capabilities that can not be inferred
  (name: string, ...args: any[]): ICap;
}

interface ICapability_defpact_test_func {
  /**
  * this is defcap doc
  */
  (
    capabilityName: \\"user.test-module.test-cap\\", 
    name: string): ICap,
}

declare module '@kadena/client' {
  interface IPactModules {
    /**
    this is module doc    
    */
    \\"user.test-module\\": {
    

      \\"defpact\\":{
        /**
        * this is defpact doc
        */
        \\"test-func\\": (
          parameterone: any,
          parametertwo: any) => string & { capability : ICapability_defpact_test_func & ICommonCapabilities} 
      }
    }
  }
}"
`;

exports[`generateDts > return type definition file for a module 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICommonCapabilities {
  (name: 'coin.GAS'): ICap;
  // let users use any other capabilities that can not be inferred
  (name: string, ...args: any[]): ICap;
}

interface ICapability_test_func {
  
  (
    capabilityName: \\"user.test-module.test-cap\\", 
    name: string): ICap,
}

declare module '@kadena/client' {
  interface IPactModules {
    
    \\"user.test-module\\": {
      \\"test-func\\": (
        parameterone: string | PactReference,
        parametertwo: boolean | PactReference) => string & { capability : ICapability_test_func & ICommonCapabilities} 

    }
  }
}"
`;

exports[`generateDts > should generate the function also if return type is a list 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICommonCapabilities {
  (name: 'coin.GAS'): ICap;
  // let users use any other capabilities that can not be inferred
  (name: string, ...args: any[]): ICap;
}

declare module '@kadena/client' {
  interface IPactModules {
    
    \\"user.test-module\\": {
      /**
      * Get all events
      */
      \\"get-events-list\\": () => string & { capability : ICommonCapabilities} 

    }
  }
}"
`;

exports[`generateDts > use any type if function parameters dont have a type 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICommonCapabilities {
  (name: 'coin.GAS'): ICap;
  // let users use any other capabilities that can not be inferred
  (name: string, ...args: any[]): ICap;
}

interface ICapability_test_func {
  
  (
    capabilityName: \\"user.test-module.test-cap\\", 
    name: string): ICap,
}

declare module '@kadena/client' {
  interface IPactModules {
    
    \\"user.test-module\\": {
      \\"test-func\\": (
        parameterone: any,
        parametertwo: boolean | PactReference) => string & { capability : ICapability_test_func & ICommonCapabilities} 

    }
  }
}"
`;

exports[`generateDts > use object type if function parameters is object{schema} 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICommonCapabilities {
  (name: 'coin.GAS'): ICap;
  // let users use any other capabilities that can not be inferred
  (name: string, ...args: any[]): ICap;
}

interface ICapability_test_func {
  
  (
    capabilityName: \\"user.test-module.test-cap\\", 
    name: string): ICap,
}

declare module '@kadena/client' {
  interface IPactModules {
    
    \\"user.test-module\\": {
      \\"test-func\\": (
        parameterone: object | PactReference,
        parametertwo: boolean | PactReference) => string & { capability : ICapability_test_func & ICommonCapabilities} 

    }
  }
}"
`;

exports[`generateDts > uses object as type when the capability accepts guard 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICommonCapabilities {
  (name: 'coin.GAS'): ICap;
  // let users use any other capabilities that can not be inferred
  (name: string, ...args: any[]): ICap;
}

interface ICapability_defpact_test_func {
  /**
  * this is defcap doc
  */
  (
    _capabilityName: \\"user.test-module.test-cap\\", 
    capabilityName: string, 
    guard: string | object): ICap,
}

declare module '@kadena/client' {
  interface IPactModules {
    /**
    this is module doc    
    */
    \\"user.test-module\\": {
    

      \\"defpact\\":{
        /**
        * this is defpact doc
        */
        \\"test-func\\": (
          parameterone: string | PactReference,
          guard: PactReference) => string & { capability : ICapability_defpact_test_func & ICommonCapabilities} 
      }
    }
  }
}"
`;

exports[`generateDts > uses the property type if there is no mapped value for that 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICommonCapabilities {
  (name: 'coin.GAS'): ICap;
  // let users use any other capabilities that can not be inferred
  (name: string, ...args: any[]): ICap;
}

declare module '@kadena/client' {
  interface IPactModules {
    
    \\"user.test-module\\": {
      \\"test\\": (param: any) => string & { capability : ICommonCapabilities} 

    }
  }
}"
`;
